### PR TITLE
【DSA】add dsa tilelang support

### PR DESCRIPTION
**Pull request by starcrown001**
### PR Category
Performance Optimization

### PR Types
New features

### Description
优化小算子性能，支持tilelang fuse版本的 DSA 实现

### 是否引起精度变化
否

### CODE REVIEW

File: src/paddlefleet/transformer/dsa_attention.py [hunk 1]
```diff
@@ -1386,49 +1788,173 @@ def forward(
...
+                topk_indices_thd = TileLangDSAFusedFunction._last_topk_indices
+                topk_indices = dsa_thd_to_bshd(topk_indices_thd, b, sq)
+
+                return core_attn_out
+            else:
+                # Fallback: original FusedDSAIndexerLoss
+                # Non-tilelang path needs softmax_scale in weights (tilelang applies it internally)
+                weights_idx = weights_idx * self.indexer.softmax_scale
+                # Convert to seq-first for FusedDSAIndexerLoss
+                q_idx_sf = q_idx.transpose(
```
[the remark this answers is not in the document]
Comment (starcrown001) [reply src/paddlefleet/transformer/dsa_attention.py]: 已修复

File: src/paddlefleet/transformer/dsa_attention.py [hunk 2]
```diff
@@ -1386,49 +1788,173 @@ def forward(
...
+                from paddlefleet.tilelang_ops.indexer.dsa_indexer import (
+                    dsa_bshd_to_thd,
+                    dsa_thd_to_bshd,
+                )
+
+                offsets = _resolve_dsa_offsets(
+                    packed_seq_params, attn_mask_startend_row_indices, b, sq
+                )
+                sparse_loss = bool(self.dsa_indexer_use_sparse_loss)
+                attn_topk = min(int(self.indexer.index_topk), sq)
```
[the remark this answers is not in the document]
Comment (starcrown001) [reply src/paddlefleet/transformer/dsa_attention.py]: 已修复

File: tests/single_card_tests/transformer/test_dsa_attention_tilelang.py [hunk 3]
```diff
@@ -0,0 +1,974 @@
...
+    and paddle.device.cuda.device_count() > 0,
+    "TileLang DSA kernels require CUDA",
+)
+class TestTileLangDSAFusedFunction(unittest.TestCase):
+    """Test TileLangDSAFusedFunction forward and backward."""
+
+    def setUp(self):
+        paddle.set_device("gpu")
+        try:
+            from paddlefleet.transformer.dsa_attention_new import (
```
[the remark this answers is not in the document]
Comment (starcrown001) [reply tests/single_card_tests/transformer/test_dsa_attention_tilelang.py]: 已修复

File: src/paddlefleet/transformer/dsa_attention.py [hunk 4]
```diff
@@ -1075,6 +1125,379 @@ def set_loss_scale(scale: Tensor):
...
+def _resolve_dsa_offsets(packed_seq_params, startend_row_indices, b, sq):
+    """Build THD offsets for TileLang varlen kernels."""
+    if (
+        packed_seq_params is not None
+        and getattr(packed_seq_params, "qkv_format", None) == "thd"
+    ):
+        return (
+            getattr(packed_seq_params, "cu_seqlens_q_padded", None)
+            or packed_seq_params.cu_seqlens_q
+        ).cast("int32")
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsa_attention.py]: P1 **优先级：P1**

这里用 Python `or` 选择 `cu_seqlens_q_padded` 会先对 `Tensor` 做真值判断；当 packed THD 路径提供 padded offsets 时，这会触发 Tensor bool 转换错误，TileLang DSA 分支在进入 kernel 前就失败。请显式判断 `None`，只在字段不存在时回退到 `cu_seqlens_q`。

```suggestion
        offsets = getattr(packed_seq_params, "cu_seqlens_q_padded", None)
        if offsets is None:
            offsets = packed_seq_params.cu_seqlens_q
        return offsets.cast("int32")
```

File: src/paddlefleet/transformer/dsa_attention.py [hunk 5]
```diff
@@ -1386,49 +1815,167 @@ def forward(
...
+                    attn_topk if sparse_loss else _pad_topk_for_tilelang(sq)
+                )
+
+                query_thd = dsa_bshd_to_thd(q_absorbed)
+                key_thd = dsa_bshd_to_thd(
+                    _make_dsa_fused_mla_key(kv_compressed, k_pos_emb)
+                )
+
+                # Single fused PyLayer: indexer topk + sparse MLA fwd + bwd
+                o_thd = TileLangDSAFusedFunction.apply(
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsa_attention.py:1910]: P1 **优先级：P1**

这个 TileLang fused 分支没有把上面构造好的 `indexer_float_mask` 或通用 `attention_mask` 传给 indexer/topk 和 sparse MLA；只依赖 offsets 能处理 packed doc 边界，但普通 padding mask、显式 attention mask 以及非 packed 的屏蔽位置都会被忽略。结果是 `dsa_tilelang_enable=True` 时可能选中并 attend 到被 mask 的 token，和现有 unfused 路径语义不一致。请在进入该分支前限制支持条件，或把 mask 语义传入 TileLang topk/attention kernel；例如先在不支持通用 mask 时回退：

```python
use_tilelang = (
    self._use_tilelang
    and attention_mask is None
    and q_absorbed is not None
    and kv_compressed is not None
    and k_pos_emb is not None
    and self._v_channels is not None
)
```

如果需要支持 padding mask，则 TileLang 的 indexer topk 和 sparse attention 都需要按同一个 mask 过滤。

File: src/paddlefleet/transformer/dsa_attention.py [hunk 6]
```diff
@@ -1075,6 +1125,379 @@ def set_loss_scale(scale: Tensor):
...
+                index_k,
+                attn_score,
+                index_score,
+                loss_topk_indices,
+                offsets,
+                token_indices,
+            )
+            dindex_q = dindex_q * ctx.loss_coeff
+            dweights = dweights * ctx.loss_coeff
+            dindex_k = dindex_k * ctx.loss_coeff
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsa_attention.py]: P1 **优先级：P1**

这里仍只按 `loss_coeff` 缩放 indexer KL 梯度，没有应用主 loss 的 backward scale。非 TileLang 路径通过 `DSAIndexerLossAutoScaler.backward` 把 `_main_loss_backward_scale` 传进 `FusedDSAIndexerLoss.backward`，CSA 的 TileLang 路径也读取同一个 scale；而这个 fused PyLayer 把 indexer loss 藏在 attention output 的 backward 里，indexer 梯度不依赖 `do`，所以梯度累积或多 microbatch loss scale 不是 1 时会和现有路径不一致。请读取 `DSAIndexerLossAutoScaler._main_loss_backward_scale` 并一起乘到三路 indexer 梯度，同时补一个设置 scale 后的梯度缩放测试。

```suggestion
            scale = DSAIndexerLossAutoScaler._main_loss_backward_scale
            if scale is None:
                scale = paddle.ones([1], dtype=dindex_q.dtype)
            coeff = ctx.loss_coeff * scale
            dindex_q = dindex_q * coeff
            dweights = dweights * coeff
            dindex_k = dindex_k * coeff
```

File: src/paddlefleet/transformer/dsa_attention.py [hunk 7]
```diff
@@ -1075,6 +1125,379 @@ def set_loss_scale(scale: Tensor):
...
+                    token_indices,
+                    use_full_loss=False,
+                )
+            )
+            loss_topk_indices = attn_topk_indices
+
+        _, lse = dsa_sparse_mla_fwd_interface(
+            query_thd,
+            key_thd,
+            attn_topk_indices.unsqueeze(1),
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsa_attention.py]: P1 **优先级：P1**

`TileLangDSAIndexerLoss.forward` 的 full-loss 分支仍先用 `attn_topk_indices` 计算 `lse`，随后用这个 sparse normalizer 去归一化 `loss_topk_indices`。当 `use_full_loss=True` 且 `loss_topk > attention_topk` 时，KL target 仍不是 full loss set 上的 softmax；这和上面 `TileLangDSAFusedFunction.backward` 新增的 `lse_full` 修复不一致。请在 full-loss 分支用 `loss_topk_indices` 重新计算 `lse`，并补一个 `attention_topk < loss_topk` 的 `TileLangDSAIndexerLoss` vs reference 对齐测试。

```python
lse_indices = loss_topk_indices if use_full_loss and loss_coeff != 0.0 else attn_topk_indices
_, lse = dsa_sparse_mla_fwd_interface(
    query_thd,
    key_thd,
    lse_indices.unsqueeze(1),
    offsets,
    token_indices,
    sm_scale=softmax_scale,
    d_v=v_channels,
)
```

File: src/paddlefleet/transformer/multi_latent_attention.py [hunk 8]
```diff
@@ -488,12 +488,16 @@ def forward(
         layer_idx = kwargs.get("layer_idx")
         use_cache = kwargs.get("use_cache", False)
 
+        needs_absorbed_q = getattr(self.config, "dsa_index_n_heads", None) is not None
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/multi_latent_attention.py]: P2 **优先级：P2**

这里把条件放宽到只要配置了 DSA 就计算 `_compute_absorbed_q`，但 `dsa_tilelang_enable` 默认是 `False`，普通 DSA/unfused 路径和 `DotProductAttention` 都不会使用 `q_absorbed`/`wv_b`。这样会让所有启用 DSA 的默认训练额外做一次 `q_nope @ W_k_b` 和相关 reshape/einsum，带来不必要的性能回退。请把条件收窄到 TileLang DSA 分支实际可能消费这些张量时再计算，例如：

```suggestion
        needs_absorbed_q = bool(
            getattr(self.config, "dsa_tilelang_enable", False)
            and getattr(self.config, "dsa_index_n_heads", None) is not None
        )
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。`dsa_tilelang_enable` 配置入口和主 fused backward 的 full-loss LSE 复算已有改进，但 TileLang DSA 路径仍有几处会影响 packed offsets、mask 语义和 indexer loss 梯度缩放的阻断问题；具体修复建议已放在行级评论中。

另外，默认未启用 TileLang 时也会额外计算 absorbed query，建议一并收窄条件以避免普通 DSA 路径性能回退。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dsa_attention.py [hunk 9]
```diff
@@ -1075,6 +1125,382 @@ def set_loss_scale(scale: Tensor):
...
+
+        is_boundary = paddle.zeros([sq], dtype="bool")
+        is_boundary[0] = True
+        is_boundary[1:] = (positions[1:] == mask[:-1]) & (mask[1:] != mask[:-1])
+
+        boundary_positions = paddle.nonzero(is_boundary).flatten().cast("int32")
+        return paddle.concat(
+            [boundary_positions, paddle.to_tensor([sq], dtype="int32")]
+        )
+    return (paddle.arange(0, b + 1, dtype="int32") * sq).cast("int32")
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsa_attention.py:1179]: P1 **优先级：P1**

当前 `_resolve_dsa_offsets` 不再读取 `packed_seq_params`。`MultiLatentAttention.get_query_key_value_tensors` 在 `packed_seq_params.qkv_format == "thd"` 时已经按 `cu_seqlens_q_padded` / `cu_seqlens_q` 处理 RoPE 和 packed 边界；但这里没有 `attn_mask_startend_row_indices` 时会返回均匀的 `[0, sq, 2*sq...]` offsets，TileLang topk/attention 会把真实 packed 样本边界切错，可能跨样本选 key 或 attend。请在固定长度 fallback 前恢复 THD cu_seqlens 分支，并继续显式判断 `None`，避免再用 Tensor `or`。

```suggestion
    if (
        packed_seq_params is not None
        and getattr(packed_seq_params, "qkv_format", None) == "thd"
    ):
        offsets = getattr(packed_seq_params, "cu_seqlens_q_padded", None)
        if offsets is None:
            offsets = packed_seq_params.cu_seqlens_q
        return offsets.cast("int32")
    return (paddle.arange(0, b + 1, dtype="int32") * sq).cast("int32")
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。此前关于 indexer loss scale、full-loss LSE 和 absorbed query 条件的意见在当前代码中看起来已处理；通用 `attention_mask` 语义的问题仍在既有行级线程中。

此次又发现 packed THD offsets 的修复不完整，具体建议已放在新的行级评论里。CI 检查目前通过；GitHub mergeability 仍显示存在合并冲突。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: src/paddlefleet/transformer/dsa_attention.py [hunk 10]
```diff
@@ -1075,6 +1136,382 @@ def set_loss_scale(scale: Tensor):
...
+    return topk
+
+
+def _resolve_dsa_offsets(packed_seq_params, startend_row_indices, b, sq):
+    """Build THD cu_seqlens offsets for TileLang varlen kernels."""
+    if startend_row_indices is not None:
+        assert b == 1, (
+            f"TileLang DSA varlen only supports batch_size=1, got {b}"
+        )
+        mask = startend_row_indices.flatten().cast("int64")
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/dsa_attention.py:1158]: P1 **优先级：P1**

这里仍然直接 `flatten()` 原始 `startend_row_indices`。本次提交让 `build_dsa_varlen_mask` 支持了 FlashMask 的 `[B, 1, S, 4]` 格式，但 TileLang 分支随后调用 `_resolve_dsa_offsets(..., attn_mask_startend_row_indices, ...)` 时传入的还是原始四列张量；`mask[:-1]` 长度会变成 `4*S-1`，和 `positions[1:]` 的 `S-1` 对不上，packed THD + TileLang 会在进入 kernel 前报 shape/broadcast 错误。请在 offsets helper 里也先归一化成 `[B, S]` 的 end-boundary，再生成 offsets，并补一个 `[1, 1, S, 4]` 走 TileLang offsets 的回归测试。

```suggestion
    if startend_row_indices is not None:
        assert b == 1, (
            f"TileLang DSA varlen only supports batch_size=1, got {b}"
        )
        indices = startend_row_indices
        if (
            indices.ndim == 4
            and indices.shape[1] == 1
            and indices.shape[3] in [1, 2, 4]
        ):
            indices = indices[:, 0, :, 0]
        elif indices.ndim != 2:
            raise ValueError(
                "attn_mask_startend_row_indices should be [B, S] or "
                "[B, 1, S, 1/2/4]"
            )
        mask = indices.reshape([b, sq]).flatten().cast("int64")
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。sequence_parallel 路径已改为显式不支持，部分 packed mask 解析也有推进；但本轮四列 FlashMask 格式还没有同步到 TileLang offsets helper，新的行级评论里有具体修复建议。

另外，既有线程中关于通用 `attention_mask` 仍会被 TileLang fused 分支绕过、以及 packed THD 无 `attn_mask_startend_row_indices` 时 offsets fallback 的问题，在当前代码里仍未看到完整修复，我没有重复开新线程。

CI 检查当前通过；GitHub mergeability 仍显示合并受阻。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — COMMENTED** [review]
已复查最新提交。此前关于 packed THD 在没有 `attn_mask_startend_row_indices` 时 offsets fallback 的问题，当前代码已恢复 `packed_seq_params.cu_seqlens_q_padded` / `cu_seqlens_q` 分支，看起来已解决；测试文件里的 `sys.path` 注入也已移除。

既有线程中关于 TileLang fused 分支忽略通用 `attention_mask`、以及多列 `attn_mask_startend_row_indices` 在 `_resolve_dsa_offsets` 中仍直接 `flatten()` 的问题，在当前代码中仍需要继续处理；本次没有新增行级评论。CI 当前通过，合并状态仍显示受冲突和审批影响。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
